content/blog/2022-04-26-e2e-cv-pt1.md

```diff
@@ -0,0 +1,331 @@
+               **  *                         *             *
+          +------------+                     *             *
+          | data_split |***                  *             *
+          +------------+   ***               *             *
+                  *            ***           *             *
+                  *               ***        *             *
+                  *                   **     *             *
+                  **                    +-------+        ***
+                    ***                 | train |     ***
+                        ***             +-------+   ***
+                            ***        **        ***
+                               ***   **      ***
+                                   **    ***
+                                  +----------+
+                                  | evaluate |
+                                  +----------+
+```
+
+The entire pipeline can be easily reproduced with the `dvc exp run` command:
+
+```bash
+$ dvc exp run
+Running stage 'check_packages':
+> python src/stages/check_pkgs.py --config=params.yaml
+...
+Running stage 'data_load':
+> python src/stages/data_load.py --config=params.yaml
+...
+Running stage 'data_split':
+> python src/stages/data_split.py --config=params.yaml
+...
+Running stage 'train':
+> python src/stages/train.py --config=params.yaml
+...
+Running stage 'evaluate':
+> python src/stages/eval.py --config=params.yaml
+...
+```
+
+Our DVC pipeline relies on the parameters defined in the
+[`params.yaml`](https://github.com/iterative/magnetic-tiles-defect/blob/main/params.yaml)
+file. By loading its contents in each stage, we can avoid hard-coded parameters.
+It also allows rerunning the whole or parts of our pipeline under a different
+set of parameters. The DVC pipeline YAML file `dvc.yaml` supports a
+[templating format](https://dvc.org/doc/user-guide/project-structure/pipelines-files#templating)
+to insert values from different sources in the YAML structure itself.
+
+DVC tracks which stages of the pipeline experienced changes and only reruns
+those. By changes, we mean _everything_ that might affect the predictive
+performance of your model i.e. changes to the dataset, source code, and/or
+parameters. This not only ensures complete reproducibility but often
+significantly reduces the time needed to rerun the whole pipeline while ensuring
+consistent results on every rerun. For example, if we only change the code in
+the evaluation stage (e.g. change what type of metrics we want to compute),
+there is no reason for us to rerun the often time-consuming data preprocessing
+and model training stages:
+
+```bash
+$ dvc exp run
+Running stage 'check_packages':
+> pipenv run pip freeze > requirements.txt
+Stage 'data_load' didn't change, skipping
+Stage 'data_split' didn't change, skipping
+Stage 'train' didn't change, skipping
+Running stage 'evaluate':
+> python src/stages/eval.py --config=params.yaml
+...
+```
+
+## Summary
```

Comment:
I think it would be helpful to have a similar summary near the beginning of the article to explain what to expect from parts 1, 2, and 3.